### PR TITLE
Fix misc compiler warnings

### PR DESCRIPTION
**Pull request by imtiazdc**
Visual Studio 2019 compiler shows multiple warnings in ZFSin. Some of them are critical and can cause hard-to-debug crashes at run time. It is a good idea to fix those proactively.

Take this for example:
...\ZFSin\ZFSin\zfs\module\zfs\zfs_vnops_windows.c: warning C4133: 'function': incompatible types - from 'QUERY_ON_CREATE_FILE_STAT_INFORMATION *' to 'FILE_STAT_INFORMATION *'

The two incompatible structs are defined as below:
C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\km\ntifs.h
typedef struct _FILE_STAT_INFORMATION {
    LARGE_INTEGER FileId;
    LARGE_INTEGER CreationTime;
    LARGE_INTEGER LastAccessTime;
    LARGE_INTEGER LastWriteTime;
    LARGE_INTEGER ChangeTime;
    LARGE_INTEGER AllocationSize;
    LARGE_INTEGER EndOfFile;
    ULONG FileAttributes;
    ULONG ReparseTag;
    ULONG NumberOfLinks;
    **ACCESS_MASK EffectiveAccess;**
} FILE_STAT_INFORMATION, *PFILE_STAT_INFORMATION;

C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\km\ntifs.h
typedef struct _QUERY_ON_CREATE_FILE_STAT_INFORMATION {
    LARGE_INTEGER FileId;
    LARGE_INTEGER CreationTime;
    LARGE_INTEGER LastAccessTime;
    LARGE_INTEGER LastWriteTime;
    LARGE_INTEGER ChangeTime;
    LARGE_INTEGER AllocationSize;
    LARGE_INTEGER EndOfFile;
    ULONG FileAttributes;
    ULONG ReparseTag;
    ULONG NumberOfLinks;
} QUERY_ON_CREATE_FILE_STAT_INFORMATION, *PQUERY_ON_CREATE_FILE_STAT_INFORMATION;

We pass the address of incompatible (smaller) data structure (**PQUERY_ON_CREATE_FILE_STAT_INFORMATION**) to _**_file_stat_information_**_ function which expects the larger version (**PFILE_STAT_INFORMATION**).
https://github.com/openzfsonwindows/ZFSin/blob/master/ZFSin/zfs/module/zfs/zfs_vnops_windows.c#L1324
**file_stat_information(IrpSp->DeviceObject, Irp, IrpSp,
              &qocContext->StatInformation);**

_**_file_stat_information_**_ goes ahead and writes to the missing field corrupting unknown (adjacent) memory. 
https://github.com/openzfsonwindows/ZFSin/blob/master/ZFSin/zfs/module/zfs/zfs_vnops_windows_lib.c#L2833
**fsi->EffectiveAccess = GENERIC_ALL;**

### CODE REVIEW

File: ZFSin/zfs/module/zfs/zfs_ioctl.c [hunk 1]
```diff
@@ -7671,6 +7671,11 @@ zfsdev_ioctl(dev_t dev, u_long cmd, caddr_t arg, int xflag, struct proc *p)
 	nvlist_free(innvl);
 
 	//arg = Irp->UserBuffer;
+	// Assuming METHOD_NEITHER is used for Control Codes used in DeviceIoControl,
+	// the output buffer is stored in Irp->UserBuffer.
+	// In using irpSp->Parameters.DeviceIoControl.Type3InputBuffer for output buffer,
+	// we are assuming the same input buffer is used for output as well.
+	ASSERT(Irp->UserBuffer == irpSp->Parameters.DeviceIoControl.Type3InputBuffer);
```
Comment (datacore-tlaurent) [inline ZFSin/zfs/module/zfs/zfs_ioctl.c]: I would not assume that here and thus would remove the ASSERT.

File: ZFSin/zfs/module/zfs/zfs_vnops_windows.c [hunk 2]
```diff
@@ -1549,10 +1550,11 @@ NTSTATUS pnp_query_id(PDEVICE_OBJECT DeviceObject, PIRP Irp, PIO_STACK_LOCATION
 
 	zmo = (mount_t *)DeviceObject->DeviceExtension;
 
-	Irp->IoStatus.Information = (void *)ExAllocatePoolWithTag(PagedPool, zmo->bus_name.Length + sizeof(UNICODE_NULL), '!OIZ');
-	if (Irp->IoStatus.Information == NULL) return STATUS_NO_MEMORY;
+	Irp->IoStatus.Information = (ULONG_PTR) ExAllocatePoolWithTag(PagedPool, 2 * (zmo->bus_name.Length + 1), '!OIZ');
```
Comment (datacore-tlaurent) [inline ZFSin/zfs/module/zfs/zfs_vnops_windows.c]: the unicode string length field is already in bytes.  you cannot multiply it again.  remove the 2 * here and below.

File: ZFSin/zfs/module/zfs/zfs_vnops_windows.c [hunk 3]
```diff
@@ -1549,10 +1550,11 @@ NTSTATUS pnp_query_id(PDEVICE_OBJECT DeviceObject, PIRP Irp, PIO_STACK_LOCATION
...
+	if (Irp->IoStatus.Information == (ULONG_PTR) NULL) return STATUS_NO_MEMORY;
 
-	RtlCopyMemory(Irp->IoStatus.Information, zmo->bus_name.Buffer, zmo->bus_name.Length);
+	RtlCopyMemory((void*)(Irp->IoStatus.Information), zmo->bus_name.Buffer, 2 * zmo->bus_name.Length);
+	((WCHAR*)(Irp->IoStatus.Information))[zmo->bus_name.Length] = L'\0';
```
Comment (datacore-tlaurent) [inline ZFSin/zfs/module/zfs/zfs_vnops_windows.c]: if you are addressing the array in WCHAR mode then you need to divide the length by 2.  Also, what was the rationale to terminate the string with NULL? is it required?

File: ZFSin/zfs/module/zfs/zfs_vnops_windows_lib.c [hunk 4]
```diff
@@ -3378,7 +3378,7 @@ NTSTATUS ioctl_disk_get_partition_info(PDEVICE_OBJECT DeviceObject, PIRP Irp, PI
 	part->HiddenSectors = (ULONG)(1L);
 	part->RecognizedPartition = TRUE;
 	part->RewritePartition = FALSE;
-	part->PartitionType = 'ZFS';
+	part->PartitionType = 0x7F;	// 'ZFS'; // https://en.wikipedia.org/wiki/Partition_type
```
Comment (datacore-tlaurent) [inline ZFSin/zfs/module/zfs/zfs_vnops_windows_lib.c]: I think this is outside the scope of fixing a compiler warning.  also, ZFS is originally Solaris and some pages on the net are saying it should be 0xBF.  I think you should reserve that change for a different PR so you can first double check with Jorgen the best value here.

File: ZFSin/spl/module/spl/spl-kmem.c [hunk 5]
```diff
@@ -63,9 +63,9 @@
 // proxy to the machine experiencing memory pressure.
 //
 // xnu vm variables
-extern volatile unsigned int vm_page_free_wanted; // 0 by default smd
-extern unsigned int vm_page_free_min; // 3500 by default smd kern.vm_page_free_min, rarely changes
-extern volatile unsigned int vm_page_free_count; // will tend to vm_page_free_min smd
+extern volatile uint64_t vm_page_free_wanted; // 0 by default smd
```
Comment (datacore-tlaurent) [inline ZFSin/spl/module/spl/spl-kmem.c]: it seems that unsigned int for this variable will be ok. it's only when the computation is made with * that the size can go beyond unsigned int but in such case the result is saved in a different uint64 variable already.  
General comment: some files are shared with linux and Jorgen will update them once in a while so if a file is one of those then we can't change the type without understanding deeper.  did you get a compile warning for that one and the ones related below ?

File: ZFSin/zfs/module/zfs/zfs_vnops_windows_mount.c [hunk 6]
```diff
@@ -922,8 +922,8 @@ void generateVolumeNameMountpoint(wchar_t *vol_mpt)
 {
 	char GUID[50];
 	wchar_t wc_guid[50];
-	generateGUID(&GUID);
-	mbstowcs(&wc_guid, GUID, 50);
+	generateGUID(GUID);
```
Comment (datacore-tlaurent) [inline ZFSin/zfs/module/zfs/zfs_vnops_windows_mount.c]: great catch !

File: ZFSin/spl/module/spl/spl-kmem.c (hunk 5 above)
Comment (imtiazdc) [reply ZFSin/spl/module/spl/spl-kmem.c]: Yes, all data type changes in this PR are made to address compiler warnings.

File: ZFSin/zfs/module/zfs/zfs_vnops_windows.c (hunk 2 above)
Comment (imtiazdc) [reply ZFSin/zfs/module/zfs/zfs_vnops_windows.c]: I see. Thanks for catching it. For some reason I thought the length was in wchars.

File: ZFSin/zfs/module/zfs/zfs_vnops_windows.c (hunk 3 above)
Comment (imtiazdc) [reply ZFSin/zfs/module/zfs/zfs_vnops_windows.c]: This change is not required as the bus_name.Length was assumed to be in wchars not bytes. As we are allocating an extra wchar in ExAllocatePoolWithTag(), it prompted me to think we needed to NULL terminate the string.

File: ZFSin/zfs/module/zfs/zfs_vnops_windows_lib.c (hunk 4 above)
Comment (imtiazdc) [reply ZFSin/zfs/module/zfs/zfs_vnops_windows_lib.c]: OK, I will revert it.
